### PR TITLE
Bugs/1592  Resolved two bugs in BatchParallel Clustering

### PR DESCRIPTION
**Pull request by mrfh92**
## Due Diligence

- General:
    - [X]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [X] unit tests: all split configurations tested
    - [X] unit tests: multiple dtypes tested
    - [X] documentation updated where needed

## Description

Issue/s resolved: #1592 

## Changes proposed:

- output labels had int32-data type as Heat-arrays, but internally they were int64-floats in torch; this causes problems when going on with computations after clustering. Now, the torch labels are cast to int32; this is no problem as nobody will have more than int32 cluster centers. 
- torch.multinomial used for K-Means++ initialization on each MPI-process has a limit of 2^24 elements (at least on GPU). Thus, if there are more than 2^24 elements to cluster on each process, we now take a uniform subsample before doing K-Means++. 

#### Does this change modify the behaviour of other functions? If so, which?
yes / no

### CODE REVIEW

File: heat/cluster/batchparallelclustering.py [hunk 1]
```diff
@@ -19,20 +19,24 @@
...
+    if X.shape[0] <= max_samples:  # torch's multinomial is limited to 2^24 categories
+        idxs = torch.zeros(n_clusters, dtype=torch.long, device=X.device)
+        idxs[0] = torch.randint(0, X.shape[0], (1,))
+        for i in range(1, n_clusters):
+            dist = torch.cdist(X, X[idxs[:i]], p=p)
+            dist = torch.min(dist, dim=1)[0]
+            idxs[i] = torch.multinomial(dist, 1)
+        return X[idxs]
+    else:  # if X is too large for the 2^24-bound, use a random subset of X
+        idxs = torch.randint(0, X.shape[0], (max_samples,))
```
Comment (JuanPedroGHM) [inline heat/cluster/batchparallelclustering.py:38]: I would consider taking the recursion out of the code. While it probably works, it tends to lower readability.

**Review by JuanPedroGHM — CHANGES_REQUESTED** [review]
Left some comments, not sure if this is the proper way to handle the bugs.

File: heat/cluster/batchparallelclustering.py [hunk 2]
```diff
@@ -289,7 +293,7 @@ def predict(self, x: DNDarray):
 
         local_labels = _parallel_batched_kmex_predict(
             x.larray, self._cluster_centers.larray, self._p
-        )
+        ).to(torch.int32)
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/cluster/batchparallelclustering.py:296]: I also thought about this and my arguments for the chosen solution were: 
* int32 saves 50% of memory compared to int64 during further processing of the outcome of the clustering
* in theory, more than int32 cluster centers can be thought of, but in practice this is completely out of scope as the runtime of our clustering algorithms heavily depend on the number of cluster centers and also the reason for doing clustering is usually to get an insight in the structure of data by grouping them into a comparably small number of clusters.

File: heat/cluster/batchparallelclustering.py [hunk 3]
```diff
@@ -19,20 +19,24 @@
 """
 
 
-def _initialize_plus_plus(X, n_clusters, p, random_state=None):
+def _initialize_plus_plus(X, n_clusters, p, random_state=None, max_samples=2**24 - 1):
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/cluster/batchparallelclustering.py:22]: Actually, this is already hard code as this is an auxiliary function that is not made available to the user directly.

[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/cluster/batchparallelclustering.py:22]: I have added a comment in the functions description.

File: heat/cluster/batchparallelclustering.py (hunk 1 above)
Comment (mrfh92) [reply heat/cluster/batchparallelclustering.py]: I have removed the recursion.

**Review by JuanPedroGHM — APPROVED** [review]
